### ISSUE TITLE
Reading invalid memory on TPU

### ISSUE THREAD

**Issue by sjmielke** [google/jax#5355]
A simple RNNLM-esque example (based on flax's seq2seq example) seems to read wrong/unallocated memory, resulting in nondeterminism on Google Colab TPUs. Using `flax.linen.transforms.scan` and `jax.value_and_grad`, both seem necessary to get this behavior, as is the sketchy looking use of multiple LSTM cells after one another (`LMCell.__call__`)...

I defined a function that looks determinstic, triggered no errors or warnings, and acts "properly" on CPU and GPU, wrapped it in `jax.value_and_grad` (full code in notebook):
```python
class LMCell(flax.linen.Module):
    @functools.partial(
        flax.linen.transforms.scan,
        length=SEQUENCE_LENGTH-1,
        variable_broadcast='params',
        split_rngs={'params': False},
    )
    @flax.linen.compact
    def __call__(self, lstm_states):
        inputs = self.param('stuff', nn.initializers.zeros, VOCAB_SIZE)
        new_lstm_states = []
        for lstm_state, lstm in zip(lstm_states, [nn.LSTMCell() for _ in range(3)]):
            new_lstm_state, inputs = lstm(lstm_state, inputs)
            new_lstm_states.append(new_lstm_state)
        logits = inputs
        return new_lstm_states, logits

class LM(flax.linen.Module):
    @flax.linen.compact
    def __call__(self):
        lstm_states = [(jnp.zeros(HIDDEN_SIZE, jnp.float32),) * 2] * 3
        _, logits = LMCell()(lstm_states)
        return -jnp.sum(logits)

params = LM().init(
    {'params': jax.random.PRNGKey(0)},
)['params']

@jax.value_and_grad
def f(params):
    return LM().apply(
        {'params': params},
        rngs={'lstm': jax.random.PRNGKey(0)},
    )
```
I now would expect this snippet should print the same value twice:
```python
print(f(params)[0])
x = jnp.log(-jnp.ones((20000, 20000)))
print(f(params)[0])
```
...but on TPUs we get `0.0` and `nan` in the linked Colab.

I first raised this as an issue in the flax project at https://github.com/google/flax/issues/769, where @avital suggested jitting the whole thing as a workaround, suggesting that there is something going wrong with the XLA TPU execution. I have neither direct access to TPUs nor do I know the first thing about debugging them and I don't have the capacity to dive into it right now, so that's why (following @avital's and @jheek's suggestion) I am now submitting the issue here as well.

Full code to reproduce in https://colab.research.google.com/drive/18Tcz0gQp7Eride_cFbw09FNDFPVIoScv?usp=sharing .

**Comment by skye** [comment]
Thanks for this clear issue report and easily-runnable repro.

I'm gonna punt on your actual issue for now, and instead note that we recently announced a new Cloud TPU experience for JAX that I've verified doesn't have this problem! We gave a talk explaining what's new if you'd like to learn more about it: [tweet](https://twitter.com/jekbradbury/status/1337528357517291520), [video](https://t.co/dk4CchuXxo?amp=1), [slides](https://docs.google.com/presentation/d/1eBfNKT3D3lEWtcn4mkgvvitKZTU7HSn4f3fGHqYgpeA/edit?resourcekey=0-1OKIlteFVZe5VeFi_1_1mw). If you don't wanna watch the whole video, the upshot is that you ssh directly to the TPU host and run JAX code there just like you would on CPU or GPU. This differs from the current setup where JAX uses an RPC layer to connect to the TPU host from a separate client VM (it appears the bug you found is in that RPC layer). The new experience is in private alpha right now, but you can sign up [here](https://docs.google.com/forms/d/14VhYvUewMWyNxGaZPEZE04frH5kEOzSs7d0pCVQYemU/edit) and we'll give you access ASAP.

It'd be wonderful if you're willing to try out the new alpha program, which gives you a lot more flexibility when running on TPU. One disadvantage is that it's not available on Colab yet, although you can use Jupyter notebooks instead (you can see a demo of that in the video). Let me know what you think -- if the alpha program doesn't sound appealing to you we'd love to hear why, and I can also prioritize tracking down this bug in that case.

**Comment by sjmielke** [comment]
Yeah, I remember seeing y'all demo that at NeurIPS and how it looked really neat (I don't expect to orchestrate across TPUs just yet either way, so it really should be super easy...)!
Feels a little surprising that it doesn't happen there, guess maybe it is some error that gets lost through the RPCs, but I'm really not familiar with just how that all works with gRPC... anyway, I filled out the form, looking forward to give it a try! Thanks!

**Comment by hawkinsp** [comment]
This is a pretty old bug, but I'm fairly sure it's fixed at head. (I'm not 100% sure, because I had to migrate the repro to use the new Flax RNN API, and I may not have done that part correctly.)

Please reopen if it still reproduces!
